Account/FxADeviceRegistrator.swift

```diff
@@ -0,0 +1,116 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Deferred
+import Shared
+
+private let log = Logger.syncLogger
+
+// The current version of the device registration, we use this to re-register
+// devices after we update what we send on device registration.
+private let DeviceRegistrationVersion = 1
+
+public enum FxADeviceRegistrationResult {
+    case Registered
+    case Updated
+    case AlreadyRegistered
+}
+
+public enum FxADeviceRegistratorError: MaybeErrorType {
+    case AccountDeleted
+    case CurrentDeviceNotFound
+    case InvalidSession
+
+    public var description: String {
+        switch self {
+        case AccountDeleted: return "Account no longer exists."
+        case CurrentDeviceNotFound: return "Current device not found."
+        case InvalidSession: return "Session token was invalid."
+        }
+    }
+}
+
+public class FxADeviceRegistrator {
+    public static func registerOrUpdateDevice(account: FirefoxAccount, state: MarriedState) -> Deferred<Maybe<FxADeviceRegistrationResult>> {
+        if let _ = account.fxaDeviceId where account.deviceRegistrationVersion == DeviceRegistrationVersion {
+            return deferMaybe(FxADeviceRegistrationResult.AlreadyRegistered)
+        }
+
+        let client = FxAClient10(endpoint: account.configuration.authEndpointURL)
```

Comment:
Nit: I think I mentioned something similar before, but if you pass in an instance of `FxAClient10` instead of the account, you'll be able to mock the client out and unit test this function. For external uses, you can wrap the method in another function that takes in the account but does the creation there instead of inside this one.